lib/ProductOpener/PackagingStats.pm

```diff
@@ -0,0 +1,647 @@
+
+When $name is "packagings-with-weights", we store stats for weights, otherwise, we store only the number of products.
+
+=cut
+
+sub add_product_components_to_stats ($name, $packagings_stats_ref, $product_ref) {
+
+	# Go through all packaging components
+	if (not defined $product_ref->{packagings}) {
+		$product_ref->{packagings} = [];
+	}
+
+	foreach my $packaging_ref (@{$product_ref->{packagings}}) {
+		my $shape = $packaging_ref->{shape} || "en:unknown";
+		my $material = $packaging_ref->{material} || "en:unknown";
+		my $weight = $packaging_ref->{weight_measured};
+
+		my @shape_parents = gen_tags_hierarchy_taxonomy("en", "packaging_shapes", $shape);
+		my @material_parents = gen_tags_hierarchy_taxonomy("en", "packaging_materials", $material);
+
+		# We will generate stats for both shapes and shapes parents
+		my @shapes_or_shapes_parents = (["shapes", [$shape, "all"]], ["shapes_parents", [@shape_parents, "all"]]);
+
+		# We will generate stats for both materials and materials parents
+		my @materials_or_materials_parents
+			= (["materials", [$material, "all"]], ["materials_parents", [@material_parents, "all"]]);
+
+		# Go through all countries
+		foreach my $country (@{$product_ref->{countries_tags}}) {
+
+			# Go through all categories (note: the product categories already contain all parent categories)
+			foreach my $category (@{$product_ref->{categories_tags}}) {
+
+				deep_val($packagings_stats_ref, ("countries", $country, "categories", $category, "n")) += 1;
+
+				# Compute stats for shapes + shapes parents
+				foreach my $shapes_or_shapes_parents_ref (@shapes_or_shapes_parents) {
+					my ($shapes_or_shapes_parents, $shapes_ref) = @$shapes_or_shapes_parents_ref;
+
+					foreach my $shape_value (@$shapes_ref) {
+
+						deep_val(
+							$packagings_stats_ref,
+							(
+								"countries", $country, "categories", $category,
+								$shapes_or_shapes_parents, $shape_value, "n"
+							)
+						) += 1;
+
+						# Compute stats for materials + materials parents
+						foreach my $materials_or_materials_parents_ref (@materials_or_materials_parents) {
+							my ($materials_or_materials_parents, $materials_ref) = @$materials_or_materials_parents_ref;
+
+							foreach my $material_value (@$materials_ref) {
+
+								deep_val(
+									$packagings_stats_ref,
+									(
+										"countries", $country,
+										"categories", $category,
+										$shapes_or_shapes_parents, $shape_value,
+										$materials_or_materials_parents, $material_value,
+										"n"
+									)
+								) += 1;
+								if (($name eq "packagings-with-weights") and (defined $weight)) {
+									deep_val(
+										$packagings_stats_ref,
+										(
+											"countries", $country,
+											"categories", $category,
+											$shapes_or_shapes_parents, $shape_value,
+											$materials_or_materials_parents, $material_value,
+											"weights", "values"
+										)
+									) .= $weight . ',';
```

Comment:
@alexgarel you are right, I did it that way because there's no deep_push() function, but in fact we don't need it, with autovivification we can just push to an array even if it does not exist yet.